### DISCUSSION TITLE
Multiple Database connection based on Request

### DISCUSSION THREAD

**Discussion by johnmcclanes001** [feathersjs/feathers#2899] [Q&A]
Right Now we have application that runs well on PHP/Mysql server. 
We are planning to move whole project to Node/Mysql and for that we have chosen the feather js as it fits very well.

I am exploring the feather for my below requirement.

We are working on accounting system.
Our client will have many companies and each company will have separate database.

User will be login from SSO system(in featherjs) and moved to default company.
and all the data will be modified in that companies database.

User will be also able to switch to companies at anytime without logged out.

Is above requirement can be fulfilled? If it possible can you please give me some idea how can I do it.
I am new to Feather JS

**Comment by daffl** [comment]
In [v5](https://dove.feathersjs.com/) you can set `params.Model` for every request to the database connection that the authenticated user belongs to. This is probably less difficult than it sounds (but depends on the database you are using). At application start you initialise the connections for all organisations

```js
const connectionSettings = app.get('connections')
const connections = {
  company1: knex(connectionSettings.company1),
  company2: knex(connectionSettings.company2),
}
app.set('connections', connections)
```

Then use a hook that sets `params.Model` after the user has been authenticated:

```js
const setCompanyModel = async context => {
  const connections = context.app.get('connections')
  const { companyName } = context.params.user
  const Model = connections[companyName]

  context.params = {
    ...context.params,
    Model
  }
}

app.service('messages').hooks({
  before: {
    all: [
      authenticate('jwt'),
      setCompanyModel
    ]
  }
})
```

There is nothing particularly fancy needed as long as you understand the Feathers basics of how hooks, services and database adapters work.

**Comment by johnmcclanes001** [comment]
I appreciate your reply and effort to help each query.
It looks good and very easy, in you solution I have to configure predefined db connection.
in my application company can be created by client and so it can not be set as predefined in code, database name will be save in one of database table and that is challenging part.


Apart from this I am trying to read database error but I did not get extractable knexjs error object.
can you please guide me.
